hca/loader.py

```diff
@@ -0,0 +1,100 @@
+from ingest.api.ingestapi import IngestApi
+from submission_broker.submission.entity import Entity
+
+from hca.submission import HcaSubmission, HandleCollision
+
+
+class HcaLoader:
+    def __init__(self, ingest: IngestApi):
+        self.ingest = ingest
+
+    def get_project(self, project_uuid: str) -> HcaSubmission:
+        hca_submission = HcaSubmission(HandleCollision.OVERWRITE)
+        project_type = 'projects'
+        project = self.__map_entity(hca_submission, project_type, project_uuid)
+        submission_type = 'submissionEnvelopes'
+        self.__map_related_entities(hca_submission, project, submission_type)
+        for submission_entity in hca_submission.get_entities(submission_type):
+            self.__map_submission_manifests(hca_submission, submission_entity)
+        return hca_submission
+
+    def get_submission(self, submission_uuid: str) -> HcaSubmission:
+        hca_submission = HcaSubmission(HandleCollision.OVERWRITE)
+        submission_type = 'submissionEnvelopes'
+        submission_entity = self.__map_entity(hca_submission, submission_type, submission_uuid)
+        self.__map_submission_manifests(hca_submission, submission_entity)
+        return hca_submission
+
+    def __map_submission_manifests(self, hca_submission: HcaSubmission, submission_entity: Entity):
+        manifest_type = 'bundleManifests'
+        self.__map_related_entities(hca_submission, submission_entity, manifest_type)
+        for manifest_entity in hca_submission.get_entities(manifest_type):
+            self.__map_manifest_content(hca_submission, manifest_entity)
+
+    def __map_entity(self, submission: HcaSubmission, entity_type: str, entity_uuid: str) -> Entity:
+        entity = submission.get_entity_by_uuid(entity_type, entity_uuid)
+        if not entity:
+            entity_attributes = self.ingest.get_entity_by_uuid(entity_type, entity_uuid)
+            entity = submission.map_ingest_entity(entity_attributes)
+            self.__add_related_entities(submission, entity)
+        return entity
+
+    def __add_related_entities(self, submission: HcaSubmission, entity: Entity):
+        entity_type = entity.identifier.entity_type
+        if entity_type == 'biomaterials' or entity_type == 'files':
```

Comment:
I find this a bit repetitive. I guess it would be possible to make this less DRY.

Reply:
Can you suggest a change to make it less so?
We are limited in some ways by the ingest-api.

Reply:
@amnonkhen do you have any further comments here? as this is now blocking.

Reply:
```suggestion
    def __add_related_entities(self, submission: HcaSubmission, entity: Entity):
        entity_type = entity.identifier.entity_type
        if entity_type == 'biomaterials' or entity_type == 'files':
            self.__map_link_type_to_link_names(submission, entity, link_type='processes', link_names=['inputToProcesses', 'derivedByProcesses'])
        elif entity_type == 'processes':
            self.__map_related_entities(submission, entity, 'protocols')
            self.__map_link_type_to_link_names(submission, entity, link_type='biomaterials', link_names=['inputBiomaterials', 'derivedBiomaterials'])
            self.__map_link_type_to_link_names(submission, entity, link_type='files', link_names=['inputFiles', 'derivedFiles'])

    def __map_link_type_to_link_names(self, submission: HcaSubmission, entity: Entity, related_entity_type: str, link_names: List[str]):
        for link_name in link_names:
            self.__map_related_entities(submission, entity, related_entity_type, link_name)
```

Reply:
worked to improve this commit suggestion with @amnonkhen, latest commit supersedes it.